src/psyclone/psyir/transformations/acc_update_trans.py

```diff
@@ -0,0 +1,332 @@
+            an OpenACC region.
+        :raises TransformationError: if the supplied Schedule contains any \
+            CodeBlocks and the 'allow-codeblocks' option is False.
+        '''
+        # transformations module file needs moving into the psyir
+        # hierarchy.
+        # pylint: disable=import-outside-toplevel
+        from psyclone.transformations import TransformationError
+
+        if not isinstance(node, Schedule):
+            raise TransformationError(f"Expected a Schedule but got a node of "
+                                      f"type '{type(node).__name__}'")
+
+        if node.ancestor(self._acc_regions):
+            raise TransformationError(
+                "Cannot apply the ACCUpdateTrans to nodes that are already "
+                "within an OpenACC compute region.")
+
+        super().validate(node, options)
+
+    def apply(self, node, options=None):
+        '''
+        Applies this transformation to the supplied Schedule. Identifies any
+        regions of code outside of ACC regions and adds the necessary ACC
+        update directives to ensure that the host and device copies of any
+        variables are kept up-to-date.
+
+        :param node: the Schedule that is to be transformed.
+        :type node: :py:class:`psyclone.psyir.nodes.Schedule`
+        :param options: any options to this transformation.
+        :type options: dict
+
+        '''
+        self.validate(node, options)
+
+        routine = node.ancestor(Routine, include_self=True)
+        self._routine_name = routine.name if routine else ""
+
+        # Call the routine that recursively adds updates to all Schedules
+        # within the supplied Schedule.
+        self._add_updates_to_schedule(node)
+
+    def _add_updates_to_schedule(self, sched):
+        '''
+        Recursively identify those statements that are not being executed
+        on the GPU and add any required OpenACC update directives.
+
+        :param sched: the schedule of statements to process.
+        :type sched: :py:class:`psyclone.psyir.nodes.Schedule`
+
+        '''
+        # We must walk through the Schedule and find those nodes representing
+        # contiguous regions of code that are not executed on the GPU. Any
+        # Call nodes are taken as boundaries of such regions
+        # because it may be that their bodies are executed on the GPU.
+        node_list = []
+        for child in sched.children[:]:
+            if isinstance(child, ACCEnterDataDirective):
+                continue
+            elif not child.walk(self._acc_regions + self._brk_nodes):
+                node_list.append(child)
+            else:
+                self._add_update_directives(node_list)
+                node_list.clear()
+                if isinstance(child, self._brk_nodes):
+                    # Conservatively add an update host statement just before
+                    # the Call node for fear of temporary operands.
```

Comment:
I could for pass-by-reference languages, since even though there is the risk of passing a reference to outdated data, an `acc update host` directive in the callee would always be able to update the host since the data location is preserved.  However, since `Call` nodes may represent function calls in pass-by-value languages where an `acc update host` directive in the callee would do nothing, I must guarantee the host has already been updated in the caller.
I must admit I wrote the comment thinking about Fortran alone, but I can update it to reflect what I just said?

Reply:
Done.